crates/executor/state_handler/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+            None => {
+                let mut new_key = Vec::new_in(Global);
+                new_key.extend_from_slice(&account_id.to_le_bytes());
+                if accumulator {
+                    new_key.push(ACC_SEPARATOR);
+                } else {
+                    new_key.push(KV_SEPARATOR);
+                }
+                new_key.extend_from_slice(key);
+                new_key
+            }
+        }
+    }
+}
+
+impl<S: Store> StdStateManager for StateHandler<S> {
+    fn kv_get<A: Allocator>(
+        &self,
+        account_id: AccountID,
+        scope: Option<AccountID>,
+        key: &[u8],
+        allocator: A,
+    ) -> Result<Option<Vec<u8, A>>, StdStateError> {
+        let constructed_key = Self::construct_key(account_id, scope, key, false);
+
+        match self.snapshot_state.get(&constructed_key, allocator) {
+            Some(value) => Ok(Some(value)),
+            None => Ok(None),
+        }
+    }
+
+    fn kv_set(
+        &mut self,
+        account_id: AccountID,
+        scope: Option<AccountID>,
+        key: &[u8],
+        value: &[u8],
+    ) -> Result<(), StdStateError> {
+        let constructed_key = Self::construct_key(account_id, scope, key, false);
+        let mut vec = Vec::new(); //TODO allocations occur here
+        vec.extend_from_slice(value);
+        self.snapshot_state.set(constructed_key, &vec);
+        Ok(())
+    }
+
+    fn kv_delete(
+        &mut self,
+        account_id: AccountID,
+        scope: Option<AccountID>,
+        key: &[u8],
+    ) -> Result<(), StdStateError> {
+        let constructed_key = Self::construct_key(account_id, scope, key, false);
+        self.snapshot_state.delete(&constructed_key);
+        Ok(())
+    }
+
+    fn accumulator_get(
+        &self,
+        account_id: AccountID,
+        scope: Option<AccountID>,
+        key: &[u8],
+    ) -> Result<u128, StdStateError> {
+        let constructed_key = Self::construct_key(account_id, scope, key, true);
+
+        match self.snapshot_state.get(&constructed_key, Global) {
+            Some(value) => {
+                let mut data = [0u8; 16];
+                data.copy_from_slice(&value);
+                Ok(u128::from_le_bytes(data))
+            }
+            None => Ok(0),
+        }
+    }
+
+    fn accumulator_add(
+        &mut self,
+        account_id: AccountID,
+        scope: Option<AccountID>,
+        key: &[u8],
+        value: u128,
+    ) -> Result<(), StdStateError> {
+        let constructed_key = Self::construct_key(account_id, scope, key, true);
+
+        let bz = self.snapshot_state.get(&constructed_key, Global);
+        let old_value: u128 = match bz {
+            Some(value) => {
+                let mut data = [0u8; 16];
+                data.copy_from_slice(&value);
+                let ov: u128 = u128::from_le_bytes(data);
+                ov
+            }
+            None => 0,
+        };
+
+        let new_value = old_value
+            .checked_add(value)
```

Comment:
I think we want a saturating add rather than an error. Or maybe we need a separate "lazy add" operation that is async. checked_add needs to be synchronous